experimental/tag_sampler.go

```diff
@@ -0,0 +1,143 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package experimental
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/uber/jaeger-client-go"
+)
+
+// TagMatcher describes which values TagMatchingSampler will match.
+type TagMatcher struct {
+	TagValue interface{} `json:"value"`
+	Firehose bool        `json:"firehose"`
+}
+
+// TagMatchingSampler samples traces that have spans with a particular tag value(s).
+type TagMatchingSampler struct {
+	jaeger.SamplerV2Base
+
+	tagKey             string
+	matchersByValue    map[interface{}]TagMatcher
+	samplerTagsByValue map[interface{}][]jaeger.Tag
+	undecided          jaeger.SamplingDecision
+}
+
+// NewTagMatchingSampler creates TagMatchingSampler with given tag key and matchers.
+func NewTagMatchingSampler(tagKey string, matchers []TagMatcher) *TagMatchingSampler {
+	// pre-generate sampler tags for all known tag values
+	samplerTagsByValue := make(map[interface{}][]jaeger.Tag)
+	matchersByValue := make(map[interface{}]TagMatcher)
+	for _, m := range matchers {
+		matchersByValue[m.TagValue] = m
+		samplerTagsByValue[m.TagValue] = []jaeger.Tag{
+			jaeger.NewTag("sampler.type", "TagMatchingSampler"),
+			jaeger.NewTag("sampler.param", fmt.Sprintf("%s=%v", tagKey, m.TagValue)),
+		}
+	}
+	return &TagMatchingSampler{
+		tagKey:             tagKey,
+		matchersByValue:    matchersByValue,
+		samplerTagsByValue: samplerTagsByValue,
+		undecided:          jaeger.SamplingDecision{Sample: false, Retryable: true, Tags: nil},
+	}
+}
+
+type tagMatchingSamplingStrategy struct {
+	Key      string       `json:"key"`
+	Matchers []TagMatcher `json:"matchers"`
+	// legacy format as map
+	Values map[string]TagMatcher `json:"values"`
+}
+
+// NewTagMatchingSamplerFromStrategyJSON creates the sampler from a JSON configuration of the following form:
+//
+//     {
+//       "key": "tagKey",
+//       "matchers": [
+//         {
+//           "value": "tagValue1",
+//           "firehose": true
+//         },
+//         {
+//           "value": 42,
+//           "firehose": false
+//         }
+//       ],
+//       "values": {
+//         "tagValue1": {
+//           "firehose": true
+//         },
+//         "tagValue2": {
+//           "firehose": false
+//         }
+//       }
+//     }
+//
+// Note that matchers can be specified either via "matchers" array (preferred),
+// or via "values" dictionary (legacy, only supports string values).
+//
+// When a given tag value appears multiple time, then last one in "matchers" wins,
+// and the one in "values" wins overall.
```

Comment:
Seems pretty complicated - can we simply disallow combining `matchers` and `value`?

Reply:
oh, so we made a mistake with the format in Node client (maps in wire payloads, I hate those). I would like to change it to a list of entries, which allows to support non-string values (like error=true tags).

It's probably not too late to remove the map version from Node, but needs coordinated change in the agent.